### ISSUE TITLE
Query support of structure_features properties

### ISSUE THREAD

**Issue by merkys** [Materials-Consortia/OPTIMADE#258]
COD will not be able to handle queries for `structure_features` property, at least not until the release of OPTiMaDe, and even some time afterwards. We do not store the symmetry-reconstructed structures yet, although we plan to do so in future. Would it be possible to reduce the query support requirement level from MUST to RECOMMENDED? If not, the COD will not be fully supporting OPTiMaDe specification for quite some time.

**Comment by rartino** [comment]
I must be missing something. Why do you need the symmetry-reconstructed structure to fill out `structure_features`? 

All you are supposed to need is:
- If there is disorder (i.e., if the cif defines any sites with <1 occupation)
- There are any unknown positions among the sites
- The structure need assemblies to be represented (i.e., the cif uses the sites grouping feature to indicate anything but completely separate groups)

**Comment by merkys** [comment]
Right, `disorder` and `assemblies` are not that difficult to assume. However, for `unknown_positions` we would have to parse CIFs and look through the coordinates, because currently we do not store this bit of information in the COD SQL database.

Moreover, I feel the definition of `unknown_positions` needs additional clarification. I will open a separate issue for that.

**Comment by rartino** [comment]
So, are you saying that COD is not storing the coordinates even of the *representative sites* in the database at all? Because then I can see how this is a problem.

Because for someone who stores those coordinates, queries on `unknown_positions` should be straightforward to translate into queries for `null`s in the coordinate table, right? (If my entry in the discussion in #260 makes sense.)

I've always tried to push us in the direction that any reasonable database should be able to implement OPTiMaDe without restructuring their data. Still, I'm very reluctant to the idea of queries on `structure_features` being optional; because it has a vital role in our handling of data compatibility! 

The idea was that carefully constructed end-user queries will almost always include a provision to exclude structures with items in `structure_features` that are not recognized/cannot be handled. That way it will be easier for implementations to not by mistake, e.g., get a bunch of structures that use `assemblies`, but the user misinterprets by ignoring the `assemblies` field.

I'm not sure what I think is the right way to proceed. To me it seems very doable to chew through all COD CIFs once and deduce if they contain "disorder", "unknown_positions", or "assemblies" and add those three boolean columns in a separate table (or even just keep that lookup data in-memory in your API implementation, since it is going to be small...) And this strikes me as information that seems useful also for people who search COD outside of OPTiMaDe. But, to force you to add this to your database does indeed go against the goal of not forcing database changes.

**Comment by merkys** [comment]
Yes, the COD does not store any sites in any *searchable* form. Instead, they are kept in CIF files on a disk, therefore, as of now we would need to scan through them for every request. We're planning to store the flag about the unknown positions in our SQL database (just as you have suggested), but this most likely will happen after the release of OPTiMaDe.

The COD already has flags for `disorder` and `assemblies`. However, one of the reasons of not having `unknown_positions` is #260.

**Comment by rartino** [comment]
However I think about this, to drop the MUST requirement on queries on `structure_features` would be very unfortunate. It was devised to solve a genuine end-user problem in relation to database-specific extensions and possible future extensions of OPTiMaDe: a client might not have any way of knowing what fields to look at to know for sure it correctly interprets structure data it receives (see discussions in #51 and #93). Allowing users to exclude entries based on unrecognized and unwanted members `structure_features` is the way we resolved that.

What we *could* do (but I am not super happy about it), is to divide `structure_features` into two fields:

- `structure_features` would be restricted to flag precisely ONLY things where there is an *additional* field somewhere (top level or below, e.g., it can be inside `sites`) that modifies how the structure should be interpreted so that it is INCORRECT to interpret the structure without that field. This is precisely the situation with `assemblies`. We would keep `structure_features` as MUST be queryable.

- Something like `structure_options` or `structure_flags` or `structure_tags` can then contain all other flags, e.g., `disorder`, `unknown_positions`, and we could re-introduce `unperiodic` that was dropped from #93. Because all those flags are things a client may want to select or unselect in a query, but if they cannot do that, it will at least surely encounter the data it does not understand in the "normal" structure fields (e.g., `nulls` among the coordinates.) To support queries on this property can then be dropped to RECOMMENDED. We can even allow implementations to only support queries on some of the flags, and even drop to just RECOMMENDED that the flags are at all present in the response for structures they apply to.

The only other alternative I see is to direct COD (and anyone in the same position) to implement `structure_features` the slow way. I.e., in a query that contains `structure_features`, do the query without that limitation and then slowly generate `cartesian_site_positions` for all of them, check if there are any `null` elements, and if so filter them out (or in). If a query uses up too much computational resources, there is always the possibility to reject it based on that. (In the long run, of course, it seem to make sense for COD to just dig through all your structures and index if `cartesian_site_positions` as returned from OPTiMaDe has any `null` elements, but that is up to you.)

**Comment by merkys** [comment]
> What we could do (but I am not super happy about it), is to divide `structure_features` into two fields [...]

I am also not very happy about it. But the status quo leaves the COD *not* conforming to the OPTiMaDe specification for the time being. Maybe we could make it to the publication, but we will have to solve #260 on the way.

**Comment by merkys** [comment]
While #279 is a great achievement, I fear it did not close this issue. Newly introduced structure feature `site_attachments` requires slow read-through of all the CIF files in the COD, thus the COD will not be able to support queries on it for some time.

Subscribing @sauliusg to this issue.

**Comment by sauliusg** [comment]
> While #279 is a great achievement, I fear it did not close this issue. Newly introduced structure feature `site_attachments` requires slow read-through of all the CIF files in the COD, thus the COD will not be able to support queries on it for some time.
> 
> Subscribing @sauliusg to this issue.

1. If we return cartesian coords, we read the table anyway;

2. My plans are to have a flag `attached_hydrogens` in the COD SQL database. This will be easy to do and backwards compatible, and we have several pending changes anyway. Let's discuss this in the COD group.

**Comment by merkys** [comment]
>     1. If we return cartesian coords, we read the table anyway;

My main concern is not for generating the values for `site_attachments`, but for making them queryable. This currently requires reading through all the COD CIFs for even the simplest queries on this feature.

>     2. My plans are to have a flag `attached_hydrogens` in the COD SQL database. This will be easy to do and backwards compatible, and we have several pending changes anyway. Let's discuss this in the COD group.

This would solve the problem for the COD (unless there could be site attachments other than hydrogen?)

**Comment by sauliusg** [comment]
> This would indeed solve the problem for the COD (unless there could be site attachments other than hydrogen?)

AFAIK CIF currently only supports hydrogens.

**Comment by merkys** [comment]
As @sauliusg and I have found a way to fix this on the COD side, I am closing the issue.
